File: CS/WebSite/App_Code/ObjectCollection.cs
using System;
using System.Collections.Generic;
using System.Web;
using System.Web.UI;
using DevExpress.Xpo;
using System.Collections;

/// <summary>
/// Summary description for ObjectCollction
/// </summary>
public class ObjectCollection<T> : IHierarchicalEnumerable {
    XPCollection<T> children;
    String parentProp;
    String childrenProp;

    public ObjectCollection(XPCollection<T> children, String parentProp, String childrenProp) {
        this.children = children;
        this.parentProp = parentProp;
        this.childrenProp = childrenProp;
    }

    public IHierarchyData GetHierarchyData(object enumeratedItem) {
        ObjectHierarchyData<T> obj = new ObjectHierarchyData<T>(children.Session, (T)enumeratedItem, parentProp, childrenProp);

        return obj as IHierarchyData;
    }

    public IEnumerator GetEnumerator() {
        return (children as IEnumerable<T>).GetEnumerator();
    }
}

File: VB/WebSite/App_Code/ObjectCollection.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Web
Imports System.Web.UI
Imports DevExpress.Xpo
Imports System.Collections

''' <summary>
''' Summary description for ObjectCollction
''' </summary>
Public Class ObjectCollection(Of T)
	Implements IHierarchicalEnumerable
	Private children As XPCollection(Of T)
	Private parentProp As String
	Private childrenProp As String

	Public Sub New(ByVal children As XPCollection(Of T), ByVal parentProp As String, ByVal childrenProp As String)
		Me.children = children
		Me.parentProp = parentProp
		Me.childrenProp = childrenProp
	End Sub

	Public Function GetHierarchyData(ByVal enumeratedItem As Object) As IHierarchyData Implements IHierarchicalEnumerable.GetHierarchyData
		Dim obj As New ObjectHierarchyData(Of T)(children.Session, CType(enumeratedItem, T), parentProp, childrenProp)

		Return TryCast(obj, IHierarchyData)
	End Function

	Public Function GetEnumerator() As IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
		Return (TryCast(children, IEnumerable(Of T))).GetEnumerator()
	End Function
End Class